pandas/core/arrays/arrow/array.py

```diff
@@ -1345,12 +1344,13 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
         else:
             data = self._data
 
-        modes = pc.mode(data, pc.count_distinct(data).as_py())
-        values = modes.field(0)
-        counts = modes.field(1)
-        # counts sorted descending i.e counts[0] = max
-        mask = pc.equal(counts, counts[0])
-        most_common = values.filter(mask)
+        if dropna:
+            data = data.drop_null()
```

Comment:
I don't know if you checked, but it might be more efficient to do this after the value_counts, so on `res` (assuming that `res` is a much shorter array, and so cheaper to filter)

Reply:
```
In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]:  arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
7.01 ms ± 148 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <-- drop_null before
6.93 ms ± 112 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <-- drop_null after
```

So might as well do this after as you suggested

Reply:
Looks like filtering after gives incorrect result for multi-mode tests. If filtering were to occur after, I would have to drop the NAs in `values` and then filter the `counts` where the NA were in `values`. To keep things simpler for now I'll give leave this to filter before the `value_counts`

Reply:
> If filtering were to occur after, I would have to drop the NAs in `values` and then filter the `counts` where the NA were in `values`

That would be something like:

```
if dropna:
    res = res.filter(res.field("values").is_valid())
```

to drop values based on one field of the struct, before calculating `most_common`.

So that line is a bit more complicated as calling `drop_null`, but only slightly. Now, it also doesn't seem to matter that much ;)

Reply:
Ah thanks. This passes the tests but appears slower than dropping the NAs beforehand for this example, so I think we should just drop the NAs beforehand for now.

```
In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
6.72 ms ± 45 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <- drop_nulls before
7.24 ms ± 87 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <- fllter, is_valild
```